connectors/spark/src/test/java/io/unitycatalog/spark/HttpRetryHandlerTest.java

```diff
@@ -0,0 +1,119 @@
+import static org.mockito.Mockito.when;
+
+import io.unitycatalog.spark.utils.Clock;
+import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+import java.time.Instant;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentMatchers;
+
+public class HttpRetryHandlerTest {
+  private Clock.ManualClock clock;
+
+  @BeforeEach
+  public void setUp() {
+    clock = (Clock.ManualClock) Clock.manualClock(Instant.now());
+  }
+
+  @AfterEach
+  public void tearDown() {
+    clock = null;
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void testRetrySucceedsAfterTwoFailures() throws IOException, InterruptedException {
+    ApiClientConf conf =
+        new ApiClientConf()
+            .setRequestMaxAttempts(5)
+            .setRequestInitialDelayMs(100L)
+            .setRequestMultiplier(2.0)
+            .setRequestJitterFactor(0.0); // Disable jitter
+
+    HttpClient mockClient = mock(HttpClient.class);
+    HttpRequest mockRequest =
+        HttpRequest.newBuilder().uri(URI.create("http://localhost:8080/api/test")).build();
+    HttpResponse.BodyHandler<String> bodyHandler = HttpResponse.BodyHandlers.ofString();
+
+    HttpResponse<String> response503 = createMockResponse(503, "Service Unavailable");
+    HttpResponse<String> response429 = createMockResponse(429, "Too Many Requests");
+    HttpResponse<String> response200 = createMockResponse(200, "Success");
+
+    // Configure mock to fail twice, then succeed
+    when(mockClient.send(
+            any(HttpRequest.class), ArgumentMatchers.<HttpResponse.BodyHandler<String>>any()))
+        .thenReturn(response503)
+        .thenReturn(response429)
+        .thenReturn(response200);
+
+    Instant start = clock.now();
+    HttpRetryHandler handler = new HttpRetryHandler(conf, clock);
+    HttpResponse<String> result = handler.call(mockClient, mockRequest, bodyHandler);
+
+    verify(mockClient, times(3))
+        .send(any(HttpRequest.class), ArgumentMatchers.<HttpResponse.BodyHandler<String>>any());
+
+    assertThat(result.statusCode()).isEqualTo(200);
+    assertThat(result.body()).isEqualTo("Success");
+
+    // Verify clock advanced for retries (2 retries)
+    // First retry delay: 100ms * 2^0 = 100ms
+    // Second retry delay: 100ms * 2^1 = 200ms
+    // Total: 300ms
+    Instant expectedTime = start.plusMillis(300);
+    assertThat(clock.now()).isAfterOrEqualTo(expectedTime);
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void testRetriesRecoverableException() throws IOException, InterruptedException {
+    ApiClientConf conf =
+        new ApiClientConf()
+            .setRequestMaxAttempts(2)
+            .setRequestInitialDelayMs(50L)
+            .setRequestMultiplier(1.0)
+            .setRequestJitterFactor(0.0); // Disable jitter
```

Comment:
there is not test testing jitter?

Reply:
for initial delay i see that different tests have different delay .. so across teh tests its getting tested that setting the initial delay to different things work correctly.

for multiplier, i see the previous test is setting it to 1. but the previous test is succeeding in only one attempt to multiplier does not kick in. so effectively you are testing multiple with only 1 value = 2? like.. if your internal code is hard coded to 2 and the conf is ignored, all your tests will pass right?